### PR TITLE
Move min and max(reduce all) to Aten(CPU)

### LINKED ISSUE #24728: Migrate `max` from the TH to Aten (CPU)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### LINKED ISSUE #33197: Inconsistent behavior of zeros in certain operations with memory_format=channels_last
## 🐛 Bug

There is some strange behavior of values stored in tensors with the channels_last format, where certain operations result in values inconsistent with the contiguous version, or where a zero diff is misinterpreted as non-zero. Probably vectorization related. Examples below.

Discovered while investigating strange test tolerances in:
https://github.com/pytorch/pytorch/pull/32538#issuecomment-584387763

A few operators (other than sqrt) appeared to have been affected. This was found in results from test_memory_format_operators.

cc @VitalyFedyunin 

## To Reproduce

On a linux GPU server:

```
In [2]: x = torch.randn((4, 3, 8, 8 ), device='cpu').contiguous(memory_format=torch.channels_last)
In [3]: r1=x.sqrt()
In [4]: r2=x.contiguous().sqrt()
In [5]: diff=torch.abs(r1-r2)
In [6]: diff.max()
Out[6]: tensor(nan)
In [7]: diff[diff==diff].max()
Out[7]: tensor(0.)
In [8]: diff[ (diff!=0) & (diff==diff) ]
Out[8]: tensor([])
In [9]: (diff/1).max()
Out[9]: tensor(5.3258e+34)```
```

On OSX:
```
>>> x = torch.randn((4, 3, 8, 8 ), device='cpu').contiguous(memory_format=torch.channels_last)
>>> r1=x.sqrt()
>>> r2=x.contiguous().sqrt()
>>> diff=torch.abs(r1-r2)
>>> diff.max()
tensor(2.1659)
```

## Expected behavior

All values in diff should be zero or nan. In cases where the values are zero, `diff/1` should still be zero in both environments.

## Environment

Linux Environment:
```
Collecting environment information...
PyTorch version: 1.5.0a0+7b50e76
Is debug build: Yes
CUDA used to build PyTorch: 9.2

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
CMake version: version 3.12.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.2.88
GPU models and configuration:
GPU 0: Tesla M40
GPU 1: Tesla M40

Nvidia driver version: 396.69
cuDNN version: /usr/local/cuda-9.2/targets/x86_64-linux/lib/libcudnn.so.7.4.2

Versions of relevant libraries:
[pip] numpy==1.15.1
[pip] numpydoc==0.8.0
[pip] torch==1.5.0a0+1016716
[conda] blas                      1.0                         mkl
[conda] magma-cuda92              2.3.0                         1    pytorch
[conda] mkl                       2019.0                      118
[conda] mkl-include               2019.0                      118
[conda] mkl-service               1.1.2            py37h90e4bf4_5
[conda] mkl_fft                   1.0.4            py37h4414c95_1
[conda] mkl_random                1.0.1            py37h4414c95_1
[conda] mkldnn                    0.14.0                        0    mingfeima
[conda] torch                     1.0.0a0+f0ed927           <pip>
[conda] torch                     1.4.0a0+ebc216a           <pip>
[conda] torch                     1.5.0a0+a7194e5           <pip>
[conda] torch                     1.4.0a0+ec92711           <pip>
[conda] torch                     1.2.0a0+646a7f9           <pip>
[conda] torch                     1.0.0a0+f3eb196           <pip>
[conda] torch                     1.5.0a0+1016716           <pip>
[conda] torch                     1.5.0a0+107436a           <pip>
[conda] torch                     1.1.0a0+71bdfe8           <pip>
[conda] torch                     1.4.0a0+36d17f4           <pip>
[conda] torch                     1.4.0a0+159835e           <pip>
[conda] torch                     1.0.0a0+623bf65           <pip>
[conda] torch                     1.0.0a0+5730d0c           <pip>
[conda] torch                     1.4.0a0+540b9da           <pip>
[conda] torch                     1.5.0a0+be6ffac           <pip>
[conda] torch                     1.1.0a0+0676ba0           <pip>
[conda] torch                     1.4.0a0+7692494           <pip>
[conda] torch                     1.4.0a0+927588d           <pip>
[conda] torch                     1.5.0a0+5ed92de           <pip>
[conda] torch                     1.5.0a0+6b37086           <pip>
[conda] torch                     1.0.0a0+d062069           <pip>
[conda] torch                     1.5.0a0+57e76f4           <pip>
[conda] torch                     1.5.0a0+738efa2           <pip>
[conda] torch                     1.5.0a0+770d990           <pip>
[conda] torch                     1.5.0a0+a9583c1           <pip>
[conda] torch                     1.5.0a0+0217a9d           <pip>
[conda] torch                     1.1.0a0+22babf1           <pip>
[conda] torch                     1.0.0a0+3ab3061           <pip>
[conda] torch                     1.5.0a0+a873850           <pip>
[conda] torch                     1.0.0a0+b911ca9           <pip>
[conda] torch                     1.5.0a0+b03e032           <pip>
[conda] torch                     1.0.0a0+69a4608           <pip>
[conda] torch                     1.0.0a0+8a64846           <pip>
[conda] torch                     1.2.0a0+34536e2           <pip>
[conda] torch                     1.4.0a0+cd3f05b           <pip>
[conda] torch                     1.0.0a0+983c364           <pip>
[conda] torch                     1.0.0a0+7270aaa           <pip>
[conda] torch                     1.5.0a0+7b50e76           <pip>
```

OSX Environment:
```
Collecting environment information...
PyTorch version: 1.5.0a0+e7f0b15
Is debug build: Yes
CUDA used to build PyTorch: None

OS: Mac OSX 10.15.2
GCC version: Could not collect
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.2
[pip3] numpydoc==0.8.0
[pip3] torch==1.5.0a0+e7f0b15
[pip3] torchserve==0.1.0b20191219
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.3                      199
[conda] mkl-include               2019.4                      233
[conda] mkl-service               1.1.2            py37hfbe908c_5
[conda] mkl_fft                   1.0.10           py37h5e564d8_0
[conda] mkl_random                1.0.2            py37h27c97d8_0
[conda] torch                     1.4.0a0+68e5172          pypi_0    pypi
[conda] torchserve                0.1.0b20191219           pypi_0    pypi

```

### PR DESCRIPTION
**Pull request by XiaobingSuper**
This PR is about port min and max(reduce all) to Aten.
Performance test script:
```
import torch
import timeit

torch.manual_seed(0)
#torch.set_num_threads(1)

device = "cpu"
print(f'device: {device}')
for op in ('max', 'min'):
    for dtype in ('torch.double', 'torch.float', 'torch.int16', 'torch.int32', 'torch.int64'):
        for n, t in [(20_000, 200000),
                     (200_000, 20000)]:
            print(f'a.{op}(), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit(f'a.{op}()', setup=f'import torch; a =(torch.torch.randn({n}) * 100).to({dtype})', number=t))
```
Test device: **skx-8180, 2 sockets**
Before:
```
a.max(), numel() == 20000 for 200000 times, dtype=torch.double
2.773961597122252
a.max(), numel() == 200000 for 20000 times, dtype=torch.double
2.3256353894248605
a.max(), numel() == 20000 for 200000 times, dtype=torch.float
3.800648272037506
a.max(), numel() == 200000 for 20000 times, dtype=torch.float
3.31692426931113
a.max(), numel() == 20000 for 200000 times, dtype=torch.int16
2.735901520587504
a.max(), numel() == 200000 for 20000 times, dtype=torch.int16
2.2510280115529895
a.max(), numel() == 20000 for 200000 times, dtype=torch.int32
2.723656536079943
a.max(), numel() == 200000 for 20000 times, dtype=torch.int32
2.228839812800288
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
2.703160767443478
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
2.3175809988752007
a.min(), numel() == 20000 for 200000 times, dtype=torch.double
2.820106916129589
a.min(), numel() == 200000 for 20000 times, dtype=torch.double
2.325718787498772
a.min(), numel() == 20000 for 200000 times, dtype=torch.float
3.833602518774569
a.min(), numel() == 200000 for 20000 times, dtype=torch.float
3.316444822587073
a.min(), numel() == 20000 for 200000 times, dtype=torch.int16
2.7308286419138312
a.min(), numel() == 200000 for 20000 times, dtype=torch.int16
2.198460517451167
a.min(), numel() == 20000 for 200000 times, dtype=torch.int32
2.730219766497612
a.min(), numel() == 200000 for 20000 times, dtype=torch.int32
2.2268200274556875
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
2.7342184390872717
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
2.320415544323623
```
After:
```
a.max(), numel() == 20000 for 200000 times, dtype=torch.double
1.7767417253926396
a.max(), numel() == 200000 for 20000 times, dtype=torch.double
0.550495645031333
a.max(), numel() == 20000 for 200000 times, dtype=torch.float
1.1113408291712403
a.max(), numel() == 200000 for 20000 times, dtype=torch.float
0.44446005020290613
a.max(), numel() == 20000 for 200000 times, dtype=torch.int16
0.5246349424123764
a.max(), numel() == 200000 for 20000 times, dtype=torch.int16
0.47057845536619425
a.max(), numel() == 20000 for 200000 times, dtype=torch.int32
0.6597231412306428
a.max(), numel() == 200000 for 20000 times, dtype=torch.int32
0.40366593934595585
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
1.767227927222848
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
0.6187495030462742
a.min(), numel() == 20000 for 200000 times, dtype=torch.double
1.7881382443010807
a.min(), numel() == 200000 for 20000 times, dtype=torch.double
0.5440589748322964
a.min(), numel() == 20000 for 200000 times, dtype=torch.float
1.1090848250314593
a.min(), numel() == 200000 for 20000 times, dtype=torch.float
0.4293213738128543
a.min(), numel() == 20000 for 200000 times, dtype=torch.int16
0.5207074657082558
a.min(), numel() == 200000 for 20000 times, dtype=torch.int16
0.41422136034816504
a.min(), numel() == 20000 for 200000 times, dtype=torch.int32
0.6145811947062612
a.min(), numel() == 200000 for 20000 times, dtype=torch.int32
0.4172037309035659
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
1.7397673893719912
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
0.596766366623342
```
Single thread:
Before:
```
a.max(), numel() == 20000 for 200000 times, dtype=torch.double
2.5068740313872695
a.max(), numel() == 200000 for 20000 times, dtype=torch.double
2.234461876563728
a.max(), numel() == 20000 for 200000 times, dtype=torch.float
3.5549037409946322
a.max(), numel() == 200000 for 20000 times, dtype=torch.float
3.2497852174565196
a.max(), numel() == 20000 for 200000 times, dtype=torch.int16
2.493077039718628
a.max(), numel() == 200000 for 20000 times, dtype=torch.int16
2.171935741789639
a.max(), numel() == 20000 for 200000 times, dtype=torch.int32
2.469274105504155
a.max(), numel() == 200000 for 20000 times, dtype=torch.int32
2.273881389759481
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
2.5818942049518228
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
2.2394551979377866
a.min(), numel() == 20000 for 200000 times, dtype=torch.double
2.5894540259614587
a.min(), numel() == 200000 for 20000 times, dtype=torch.double
2.331936141476035
a.min(), numel() == 20000 for 200000 times, dtype=torch.float
3.590122046880424
a.min(), numel() == 200000 for 20000 times, dtype=torch.float
3.255849950015545
a.min(), numel() == 20000 for 200000 times, dtype=torch.int16
2.5205496419221163
a.min(), numel() == 200000 for 20000 times, dtype=torch.int16
2.168218174017966
a.min(), numel() == 20000 for 200000 times, dtype=torch.int32
2.658622432500124
a.min(), numel() == 200000 for 20000 times, dtype=torch.int32
2.3376982398331165
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
2.496626536361873
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
2.2504652086645365
```
After:
```
a.max(), numel() == 20000 for 200000 times, dtype=torch.double
1.9525171788409352
a.max(), numel() == 200000 for 20000 times, dtype=torch.double
1.6108122132718563
a.max(), numel() == 20000 for 200000 times, dtype=torch.float
1.2444602297618985
a.max(), numel() == 200000 for 20000 times, dtype=torch.float
0.7705567870289087
a.max(), numel() == 20000 for 200000 times, dtype=torch.int16
0.6575072864070535
a.max(), numel() == 200000 for 20000 times, dtype=torch.int16
0.13242999743670225
a.max(), numel() == 20000 for 200000 times, dtype=torch.int32
0.829406064003706
a.max(), numel() == 200000 for 20000 times, dtype=torch.int32
0.35575105529278517
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
1.6426756298169494
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
1.4049720335751772
a.min(), numel() == 20000 for 200000 times, dtype=torch.double
2.029639278538525
a.min(), numel() == 200000 for 20000 times, dtype=torch.double
1.6363644907251
a.min(), numel() == 20000 for 200000 times, dtype=torch.float
1.3821239182725549
a.min(), numel() == 200000 for 20000 times, dtype=torch.float
0.834847847931087
a.min(), numel() == 20000 for 200000 times, dtype=torch.int16
0.6913397628813982
a.min(), numel() == 200000 for 20000 times, dtype=torch.int16
0.1370067736133933
a.min(), numel() == 20000 for 200000 times, dtype=torch.int32
0.8190992185845971
a.min(), numel() == 200000 for 20000 times, dtype=torch.int32
0.3640836915001273
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
1.6516661625355482
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
1.4111155439168215
```
Fixes: #33197

Fix https://github.com/pytorch/pytorch/issues/24728, https://github.com/pytorch/pytorch/issues/24729

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin , @ngimel

**Comment by VitalyFedyunin** [comment]
Can you please run additional single threaded benchmarks. Also `int64` perf degradation looks suspicious.  Might be because there is no such thing as NaN in ints and you can improve code dropping additional if.

File: aten/src/ATen/native/cpu/zmath.h [hunk 1]
```diff
@@ -199,32 +200,56 @@ inline std::complex<double> trunc_impl (std::complex<double> z) {
 
 template <typename TYPE>
 inline TYPE max_impl (TYPE a, TYPE b) {
-  return std::max(a, b);
+  if (_isnan<TYPE>(a) || _isnan<TYPE>(b)) {
```
Comment (xuhdev) [inline aten/src/ATen/native/cpu/zmath.h]: Why not use `std::isnan`?

Comment (XiaobingSuper) [reply aten/src/ATen/native/cpu/zmath.h]: @xuhdev ,  https://github.com/pytorch/pytorch/blob/9f7708eecb25268702ac107471fd13972fb4f5d2/aten/src/ATen/NumericUtils.h#L16 say **std::isnan isn't performant to use on integral types**.  there can use **std::isnan** as @VitalyFedyunin suggestion if not care Nan case for int.

Comment (xuhdev) [reply aten/src/ATen/native/cpu/zmath.h]: I think you are right, according to https://en.cppreference.com/w/cpp/numeric/math/isnan :

> A set of overloads or a function template accepting the arg argument of any integral type. Equivalent to (2) (the argument is cast to double).

**Comment by XiaobingSuper** [comment]
single thread performance:
Before:
```
device: cpu
a.max(), numel() == 20000 for 200000 times, dtype=torch.double
2.5958177908323705
a.max(), numel() == 200000 for 20000 times, dtype=torch.double
2.2637504609301686
a.max(), numel() == 20000 for 200000 times, dtype=torch.float
2.5624412591569126
a.max(), numel() == 200000 for 20000 times, dtype=torch.float
2.2108274307101965
a.max(), numel() == 20000 for 200000 times, dtype=torch.int16
2.546321148984134
a.max(), numel() == 200000 for 20000 times, dtype=torch.int16
2.1951667023822665
a.max(), numel() == 20000 for 200000 times, dtype=torch.int32
2.5384745127521455
a.max(), numel() == 200000 for 20000 times, dtype=torch.int32
2.2654318581335247
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
2.6368427388370037
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
2.3595943828113377
a.min(), numel() == 20000 for 200000 times, dtype=torch.double
2.6794958747923374
a.min(), numel() == 200000 for 20000 times, dtype=torch.double
2.2660013218410313
a.min(), numel() == 20000 for 200000 times, dtype=torch.float
4.375531195662916
a.min(), numel() == 200000 for 20000 times, dtype=torch.float
3.9738557641394436
a.min(), numel() == 20000 for 200000 times, dtype=torch.int16
2.6021414771676064
a.min(), numel() == 200000 for 20000 times, dtype=torch.int16
2.217277537100017
a.min(), numel() == 20000 for 200000 times, dtype=torch.int32
2.539621968753636
a.min(), numel() == 200000 for 20000 times, dtype=torch.int32
2.2727541951462626
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
2.6566950669512153
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
2.364656137768179
```
After:
```
device: cpu
a.max(), numel() == 20000 for 200000 times, dtype=torch.double
1.1517636110074818
a.max(), numel() == 200000 for 20000 times, dtype=torch.double
1.2310124468058348
a.max(), numel() == 20000 for 200000 times, dtype=torch.float
0.8446659930050373
a.max(), numel() == 200000 for 20000 times, dtype=torch.float
0.3539451779797673
a.max(), numel() == 20000 for 200000 times, dtype=torch.int16
0.6152581977657974
a.max(), numel() == 200000 for 20000 times, dtype=torch.int16
0.13027244294062257
a.max(), numel() == 20000 for 200000 times, dtype=torch.int32
0.7246702602133155
a.max(), numel() == 200000 for 20000 times, dtype=torch.int32
0.2521747397258878
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
3.588684367015958
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
3.1973710479214787
a.min(), numel() == 20000 for 200000 times, dtype=torch.double
1.1078475271351635
a.min(), numel() == 200000 for 20000 times, dtype=torch.double
1.234666217584163
a.min(), numel() == 20000 for 200000 times, dtype=torch.float
0.894020417239517
a.min(), numel() == 200000 for 20000 times, dtype=torch.float
0.3695054338313639
a.min(), numel() == 20000 for 200000 times, dtype=torch.int16
0.613672009203583
a.min(), numel() == 200000 for 20000 times, dtype=torch.int16
0.1299265082925558
a.min(), numel() == 20000 for 200000 times, dtype=torch.int32
0.7175788767635822
a.min(), numel() == 200000 for 20000 times, dtype=torch.int32
0.2461460060440004
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
3.639761713333428
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
3.1744841523468494
```

**Comment by XiaobingSuper** [comment]
I think there has a performance issue for multi_thread reduce,  the single thread can get a better performance than multi_thread(amost **~2x**) in this case. for other ops, such as **sum, prod,** they also have same problem, see:
**num_thread = 56**:
```
device: cpu
a.prod(), numel() == 20000 for 200000 times, dtype=torch.float32
0.741498313844204
a.prod(), numel() == 200000 for 20000 times, dtype=torch.float32
2.1368089811876416
a.sum(), numel() == 20000 for 200000 times, dtype=torch.float32
0.739946136251092
a.sum(), numel() == 200000 for 20000 times, dtype=torch.float32
2.097711270209402
```
**num_thread = 1:**
```
device: cpu
a.prod(), numel() == 20000 for 200000 times, dtype=torch.float32
0.7309436281211674
a.prod(), numel() == 200000 for 20000 times, dtype=torch.float32
0.2751764222048223
a.sum(), numel() == 20000 for 200000 times, dtype=torch.float32
0.7341888980008662
a.sum(), numel() == 200000 for 20000 times, dtype=torch.float32
0.2784042269922793
```
Test script:
```
import torch

 import timeit

 torch.manual_seed(1)
 torch.set_num_threads(1)

 device = "cpu"
 print(f'device: {device}')
 for op in ('prod', 'sum'):
     for dtype in ('torch.float32',):
         for n, t in [(20_000, 200000),
                      (200_000, 20000)]:
             print(f'a.{op}(), numel() == {n} for {t} times, dtype={dtype}')
             print(timeit.timeit(f'a.{op}()', setup=f'import torch; a =(torch.torch.randn({n}) * 10).to({dtype})', number=t))
```
For big input size, the single thread performance is **9x** than multi thread performance.

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin , I rewrite the code using **parallel_reduce** which can get a good performace than TensorIterator for reduce all cases,  we can now get a big performance improvement expect **int64**, see the details in the performance data.  if this is ok, I will rewrite other reduce all operators, such as sum, prod, std and var, and will get good performance than before. Thanks!

**Comment by ngimel** [comment]
sum and prod are already using vectorized TensorIterator kernels, it's unlikely that their performance can be improved. However, there's a curious case of a particular .norm configuration (#34279) with very bad performance, you might want to take a look at that.

**Comment by XiaobingSuper** [comment]
@ngimel, TensorIterator can work well for given reduce dim, but for reduce all case, the multi_thread's performance is not good enough, ther has simple implementation in [56ac269](https://github.com/XiaobingSuper/pytorch/commit/56ac2692911280a23be32c48b5b456d732810272) uisng **parallel_reduce** , for big input size, **parallel_reduce**  can get **~5x** performance improvement, see

TensorIterator:
```
device: cpu
a.prod(), numel() == 20000 for 200000 times, dtype=torch.float32
0.7129124458879232
a.prod(), numel() == 200000 for 20000 times, dtype=torch.float32
1.9596283761784434
a.sum(), numel() == 20000 for 200000 times, dtype=torch.float32
0.7430614968761802
a.sum(), numel() == 200000 for 20000 times, dtype=torch.float32
1.909514768049121
```
parallel_reduce:
```
device: cpu
a.prod(), numel() == 20000 for 200000 times, dtype=torch.float32
0.9682887392118573
a.prod(), numel() == 200000 for 20000 times, dtype=torch.float32
0.39769155997782946
a.sum(), numel() == 20000 for 200000 times, dtype=torch.float32
0.9798075156286359
a.sum(), numel() == 200000 for 20000 times, dtype=torch.float32
0.3649212485179305
```

**Comment by ngimel** [comment]
What's the status of this PR? Is it ready for review? Is int64 perf regression solved?

**Comment by XiaobingSuper** [comment]
@ngimel, there only have one problem, int64 performance regression, for int64 case, I found **_mm256_extract_epi64**  consume much time in:
https://github.com/pytorch/pytorch/blob/4167db11f76634a75754c0245d79adb18798cec1/aten/src/ATen/cpu/vec256/vec256_int.h#L516-L533
Is it normal?  BTW, even without vectorization, we can also get a good performance than before using this patch:
```
diff --git a/aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp b/aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp
index b26a514ccc..9d118b8f23 100644
--- a/aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp
+++ b/aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp
@@ -28,10 +28,16 @@ inline void reduce_all_impl(
   auto input_data = input.data_ptr<scalar_t>();
   scalar_t result = at::parallel_reduce(0, input_numel, internal::GRAIN_SIZE, ident_v,
     [&](int64_t start, int64_t end, const scalar_t ident) -> scalar_t {
+      scalar_t partial_out = ident;
+      for (int64_t i = start; i < end; i++) {
+        partial_out = op(partial_out, input_data[i]);
+      }
+      /*
       scalar_t partial_out = vec256::reduce_all<scalar_t>(
         [=](Vec x, Vec y) { return vop(x, y); },
         input_data + start,
         end - start);
+      */
       return partial_out;
     }, op);
   output.fill_(result);
diff --git a/third_party/fbgemm b/third_party/fbgemm
index af57f36db0..baab5386e2 160000
--- a/third_party/fbgemm
+++ b/third_party/fbgemm
@@ -1 +1 @@
-Subproject commit af57f36db0f283ab377843784681394daf0e6bff
+Subproject commit baab5386e2b632d054cf43796631bb0338fd8c9f
```

**Comment by ngimel** [comment]
I'm not sure - can you try a non-vectorized implementation for int64 and see if it's better?

**Comment by XiaobingSuper** [comment]
@ngimel, the non-vectorized implementation for int64 can get the best performance:

num_thread = 1
TH path:
```
device: cpu
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
2.501999666914344
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
2.252055495046079
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
2.4950898000970483
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
2.2494910173118114
```
vectorized
```
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
5.421072099357843
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
4.986506130546331
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
5.375462950207293
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
5.0889328103512526
```
non-vectorized:
```
device: cpu
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
1.9133615167811513
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
1.5558369308710098
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
1.902137327939272
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
1.5547090442851186
```

num_thread = 56
TH path:
```
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
2.506502198986709
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
2.2478088419884443
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
2.5004060305655003
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
2.2456414829939604
```
vectorized
```
device: cpu
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
5.555780376307666
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
1.2713476000353694
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
5.229991992004216
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
1.2193763088434935
```
non-vectorized:
```
device: cpu
a.min(), numel() == 20000 for 200000 times, dtype=torch.int64
1.7868667198345065
a.min(), numel() == 200000 for 20000 times, dtype=torch.int64
0.5336812091991305
a.max(), numel() == 20000 for 200000 times, dtype=torch.int64
1.7609290108084679
a.max(), numel() == 200000 for 20000 times, dtype=torch.int64
0.5401007682085037
```
I am now analysing the details reason. Thanks!

**Comment by XiaobingSuper** [comment]
@ngimel @VitalyFedyunin, it is ready for review now, thanks!

File: aten/src/ATen/native/cpu/zmath.h [hunk 2]
```diff
@@ -199,32 +200,56 @@ inline std::complex<double> trunc_impl (std::complex<double> z) {
...
+    return std::complex<double>(std::numeric_limits<double>::quiet_NaN());
+  } else {
+    return std::complex<double>(std::abs(a) > std::abs(b) ? a : b);
+  }
 }
 
 template <typename TYPE>
 inline TYPE min_impl (TYPE a, TYPE b) {
-  return std::min(a, b);
+  if (_isnan<TYPE>(a) || _isnan<TYPE>(b)) {
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/zmath.h]: It's better to have min_impl and max_impl as follows (this way you don't have to have separate instantiations for complex float and double):
```
template <typename TYPE, std::enable_if_t<!c10::is_complex_t<TYPE>::value, int> =0>
inline TYPE min_impl(TYPE a, TYPE b) {
    return std::min(a, b);
}

template <typename TYPE, std::enable_if_t<c10::is_complex_t<TYPE>::value, int> =0>
inline TYPE min_impl (TYPE a, TYPE b) {
    return (std::abs(a) < std::abs(b) ? a : b);
}
```
I'm omitting _isnan branch here for clarity, of course you should have it. Similar approach is used for `_isnan` definition, you can take a look here https://github.com/pytorch/pytorch/blob/9f7708eecb25268702ac107471fd13972fb4f5d2/aten/src/ATen/NumericUtils.h#L16

**Review by ngimel — APPROVED** [review]
Generally looks good, I have a minor suggestion for  templates for min_impl/max_impl.

File: aten/src/ATen/native/cpu/zmath.h (hunk 2 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/cpu/zmath.h]: @ngimel, it is a good point, but there has problem, for complex number , if one of them is Nan, how to return a Nan complex number? **std::numeric_limits** should given a data type, float of double.

Comment (ngimel) [reply aten/src/ATen/native/cpu/zmath.h]: This by itself is not a problem, `return TYPE(std::numeric_limits<double>::quiet_NaN())` would do the right thing. However, now I realize that it does not correctly propagate complex nans - if imaginary part of a or b is nan, a number different from a or b will be returned, with real part equal to nan, and imaginary set to 0. So the complex template should be
```
   if (_isnan(a)) 
      return a;
   else if (_isnan(b))
      return b;
   else ...
```

Comment (XiaobingSuper) [reply aten/src/ATen/native/cpu/zmath.h]: @ngimel , thanks!

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**